scipy/stats/stats.py

```diff
@@ -5718,3 +5721,195 @@ def rankdata(a, method='average'):
 
     # average method
     return .5 * (count[dense] + count[dense - 1] + 1)
+
+
+def rvs_ratio_unif(pdf, size=1, x1=None, x2=None, y2=None, shift=0,
+                   maxiter=None, bounds=None, random_state=None):
+    """
+    Generate random samples from a probability density function using the
+    ratio-of-uniforms method.
+
+    Parameters
+    ----------
+    pdf : callable
+        The probability density function of the distribution.
+    size : int or tuple of ints, optional
+        Defining number of random variates (default is 1).
+    x1 : float, optional
+        The lower bound of the bounding rectangle in the x-direction.
+        Default is None. In that case, attempt to find bound numerically.
+    x2 : float, optional
+        The upper bound of the bounding rectangle in the x-direction.
+        Default is None. In that case, attempt to find bound numerically.
+    y2 : float, optional
+        The upper bound of the bounding rectangle in the y-direction.
+        Default is None. In that case, attempt to find bound numerically.
+    shift : float, optional.
+        Shift parameter of ratio-of-uniforms method, see Notes. Default is 0.
+    maxiter : int, optional
+        Maximum number of iterations to generate the random variables.
+        Default is None. In that case, maxiter is set to a value that
+        ensures successful simulation with a high probability.
+    bounds : tuple of floats (a, b), optional
+        Default is None. In case any of the values x1, x2, y2 are not
+        specified (i.e. equal to None), attempt to find the value numerically
+        over the interval from a to b.
+    random_state : None or int or np.random.RandomState instance, optional
+        If int or RandomState, use it for drawing the random variates.
+        If None, use np.random.RandomState. Default is None.
+
+    Returns
+    -------
+    rvs : array
+        the random variates distributed according to the probability
+        distrubtion, pdf
+
+    Notes
+    -----
+    Given a univariate probability density function (pdf) f and a constant c,
+    define the set A(c) = {(x, y) : 0 < y <= sqrt(f(x/y + c)) }.
+    If (U, V) is a random vector uniformly distributed over A(c), then U/V + c
+    follows a distribution with pdf f.
+
+    The above result (see [1]_, [2]_) can be used to sample random variables
+    using only the pdf, i.e. no inversion of the cdf is required. Typical
+    choices of c are zero or the mode of f. The set A(c) is a subset of the
+    rectangle R = [-x1, x2] x [0, y2] where
```

Comment:
The interval `[-x1, x2] x [0, y2]` and the ratio of areas below `2 * y2 * (x2 - x1)` don't seem consistent to me. 

I'm looking at the example of a uniform distribution on the interval `[2, 6]` (see the Kinderman/Monahan 1977 paper listed in another comment below), using c=shift=0.  Then `y2=1/sqrt(6-2) = 0.5, x1=1, x2=3`.  The region `A(c)` appears to be the triangle with vertices `(0, 0), (0.5, 1), (0.5, 3)` and area `0.5`,  and trying to reconcile that with the rectangle R, and its implied area `y2 * (x2 - x1)=1`.  Is `R(c)=[-1, 3]x[0, 0.5]` with area 2?  Or is `R(c)=[1, 3]x[0, 0.5]` with area 1?  Certainly the latter doesn't contain `A(c)`, so I'm guessing the former?

Reply:
The calculation of `x1` in the example above isn’t quite correct. `x1` needs to be calculated over all of the reals, not just the interval [2,6], so x1 should be 0, not 0.5.  A better example might be a uniform distribution on the interval `[-1, 3]`, with `y2=0.5, x1=-0.5 and x2=1.5`.  That suggests the rectangle `R=[x1, x2] x [0, y2]`.